src/public/manageDomain.ts

```diff
@@ -0,0 +1,150 @@
+          >
+            Remove
+          </button>
+        </div>
+        <div class="invalid-feedback">
+          The variable name must contain only letters, number, and/or the underscore character.
+        </div>
+      </div>
+    `
+
+    const nameInputElement = environmentElement.querySelector(
+      '#name'
+    ) as HTMLInputElement
+    const valueInputElement = environmentElement.querySelector(
+      '#value'
+    ) as HTMLInputElement
+    nameInputElement.addEventListener('input', () => {
+      // validate the input for valid variable name with regex
+      // valid: letters, numbers and underscore
+      const name = nameInputElement.value.toUpperCase()
+      if (name.match(/^[A-Z0-9_]+$/g)) {
+        this.name = nameInputElement.value.toUpperCase()
+        this.isValid = true
+        nameInputElement.classList.remove('is-invalid')
+      } else {
+        this.isValid = false
+        nameInputElement.classList.add('is-invalid')
+      }
+    })
+    valueInputElement.addEventListener(
+      'input',
+      () => (this.value = valueInputElement.value)
+    )
+    environmentElement
+      .querySelector('#removeVariable')
+      .addEventListener('click', () => {
+        environmentElement.remove()
+        environmentVariables.splice(environmentVariables.indexOf(this), 1)
+      })
+
+    if (this.name && this.value) {
+      nameInputElement.value = this.name
+      valueInputElement.value = this.value
+      this.isValid = true
+    } else {
+      nameInputElement.classList.add('is-invalid')
+    }
+  }
+}
+
+document.getElementById('addEnvButton').onclick = (): void => {
+  environmentVariables.push(new EnvironmentItem())
+}
+
+document.getElementById('submitEnvButton').onclick = (): void => {
+  // check if there's at least one variable to submit
+  if (environmentVariables.length === 0) {
+    return alert('Error: add at least one variable before submitting.')
+  }
+  // check if all variables are valid
+  if (environmentVariables.some(env => !env.isValid)) {
+    return alert(
+      'Error: some fields are not valid, please fix them before submitting.'
+    )
+  }
+  // confirm if it's OK to restart the app and send the request
+  if (
+    confirm(
+      'Submitting will recreate the container for your app.' +
+        'All data not saved inside the main app folder will be lost. Is this OK?'
+    )
+  ) {
+    // convert to the request body format
+    // { variables: { NAME: value, ... }}
+    const body = environmentVariables.reduce((acc, { name, value }) => {
+      if (name && value) {
+        acc[name] = value
+      }
+      return acc
+    }, {})
+    fetch(`/api/mappings/${fullDomain}/environment`, {
+      method: 'PUT',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({ variables: body })
+    })
+      .then(response => (response.status !== 201 ? response.json() : {}))
+      .then((body: ContainerResponse) =>
+        body.message
```

Comment:
`.message` seems unintuitive as an error. Could we change it to `.error` or even better. Make it actually return with an error status code to be handled with `.catch` or `try...catch`?

Reply:
`message` is consistent with the remaining code base and to the docker API.

Reply:
Also, failed status code are already sent from the server. But since we have to parse the JSON to get the error message, I don't know if there's a simple way to do this inside `.catch()`.

Reply:
Unless I'm wrong, I thought any error status codes from `fetch` immediately went to the `.catch()` chain.

Reply:
Only for network errors. From the [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch):

>The Promise returned from fetch() won’t reject on HTTP error status even if the response is an HTTP 404 or 500. Instead, it will resolve normally (with ok status set to false), and it will only reject on network failure or if anything prevented the request from completing.

Reply:
Interesting ok. I wonder where I got the idea that it would reject lol
Thanks 👍